openupgrade_scripts/scripts/hr/16.0.1.1/post-migration.py

```diff
@@ -0,0 +1,167 @@
+            "hr.plan.activity.type can only be linked to one hr.plan. "
+            "Fix these data before migrating to avoid data loss.",
+            res[0],
+        )
+
+
+def m2m_to_o2m(
+    env,
+    model,
+    field,
+    source_relation_table,
+    relation_source_field,
+    relation_comodel_field,
+):
+    """Transform many2many relations into one2many (with possible data
+    loss).
+
+    Use rename_tables() in your pre-migrate script to keep the many2many
+    relation table and give them as 'source_relation_table' argument.
+    And remove foreign keys constraints with remove_tables_fks().
+
+    :param model: The target registery model
+    :param field: The field that changes from m2m to o2m
+    :param source_relation_table: The (renamed) many2many relation table
+    :param relation_source_field: The column name of the 'model' id
+    in the relation table
+    :param relation_comodel_field: The column name of the comodel id in
+    the relation table
+    """
+    columns = env[model]._fields.get(field)
+    target_table = env[columns.comodel_name]._table
+    target_field = columns.inverse_name
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE %(target_table)s AS target
+        SET %(target_field)s=source.%(relation_source_field)s
+        FROM %(source_relation_table)s AS source
+        WHERE source.%(relation_comodel_field)s=target.id
+        """,
+        {
+            "target_table": AsIs(target_table),
+            "target_field": AsIs(target_field),
+            "source_relation_table": AsIs(source_relation_table),
+            "relation_source_field": AsIs(relation_source_field),
+            "relation_comodel_field": AsIs(relation_comodel_field),
+        },
+    )
+
+
+def create_work_contact(env):
+    """Create work_contact_id for model hr.employee.base
+
+    If the employee is linked to a res.user, we set the partner_id of the
+    res.user as a work_contact_id.
+
+    If the employee is not linked to a res.user. Then we try to match an
+    existing partner with the same email address. If one is found, then
+    we assign it as work_contact_id. If several are found, then nothing
+    is done and a warning is send in the log.
+    """
+    employees = env["hr.employee"].search([])
+
+    for employee in employees:
+        if employee.user_id and employee.user_id.partner_id:
+            # FIXME: not checking that the partner has the same email
+            # and mobile than the employee may lead to a loss of data.
+            # work_email and mobile_phone will be replaced by the one
+            # from the partner.
+            employee.work_contact_id = employee.user_id.partner_id
+            _logger.info(
+                "Set work_contact_id of hr.employee(%s) to "
+                "res.partner(%s) (the res.user partner).",
+                employee.id,
+                employee.user_id.partner_id.id,
+            )
```

Comment:
@remi-filament I implemented your idea to link the partner linked to the user_id if there is one.

But I'm not sure if such a link with work_contact_id should be done if the `mobile` and `email` of the employee and the partner are not the same. Because this may lead to a loss of data.

And I think that the goal of an openupgrade script is to avoid losing data.

Should I check that the partner found on the user_id has the same `mobile` and `email` as the employee before assigning it to `work_contact_id` ?

Reply:
@remytms thanks for that !
Yes it would make sense to check this out (and maybe have some specific action in case fields are empty on partner or employee side)

Reply:
Indeed, I will address this.

Reply:
@remi-filament Done.